### PR TITLE
Limit maximum alignment in #[repr(align)] to 4K

### PR DESCRIPTION
**Pull request by Amanieu**
Let's try this on crater to see the impact.

cc #70022 #70143 #70144

### CODE REVIEW

File: src/librustc_attr/builtin.rs [hunk 1]
```diff
@@ -908,11 +908,12 @@ pub fn find_repr_attrs(sess: &ParseSess, attr: &Attribute) -> Vec<ReprAttr> {
...
                             if literal.is_power_of_two() {
-                                // rustc::ty::layout::Align restricts align to <= 2^29
-                                if *literal <= 1 << 29 {
+                                // Many targets don't support global variables
+                                // with alignment greater than the page size.
+                                if *literal <= 4096 {
                                     Ok(*literal as u32)
                                 } else {
-                                    Err("larger than 2^29")
+                                    Err("larger than 4096")
```
Comment (Centril) [inline src/librustc_attr/builtin.rs]: if we land this, then it would be good to mention what this magic number is about

Comment (Centril) [inline src/librustc_attr/builtin.rs]: Also, let's add a regression test at `(4096 + 1).next_power_of_two()`

File: src/test/ui/structs-enums/align-struct.rs [hunk 2]
```diff
@@ -67,9 +67,9 @@ struct AlignContainsPacked4C {
 
 // The align limit was originally smaller (2^15).
 // Check that it works with big numbers.
```
Comment (hanna-kruppe) [inline src/test/ui/structs-enums/align-struct.rs]: Irrelevant for crater run but this comment is now wrong.

**Comment by Centril** [comment]
@craterbot check

**Comment by chorman0773** [comment]
Limiting to 4k alignment could hurt specialized structures in an embedded context. For example, I haved used align(32k) (or rather the equivalent) to place objects in a half-bank on 65816. Half banks are specifically important, as that is conventionally the mapping division point, and the largest page size used by the mapper. It also makes it easier to place something at a precise address with fewer linker hacks.

**Comment by bjorn3** [comment]
`maligned` is the root regression in most cases.

**Comment by Amanieu** [comment]
There are 3 root regressions:
- greenthread-future:
```rust
/// This struct is allocated on heap whose start address is aligned to 0x2000.
/// So that we can quickly locate it from stack pointer (just like Linux).
```
- maligned:
```rust
//! A library for getting aligned vectors, slices, and arrays of bytes or any other type
//!
//! The `A*` structs are aligned as indicated and hold a single array of bytes of the same size
//! These implement an `Alignment` trait that can be used for always aligned byte arrays for faster
//! memory copies or copies by hardware. The arrays are accessible as slices using `as_bytes()` or
//! `as_bytes_mut()` or by dereferencing the struct.
```
- reclaim:
```rust
//! Thin wrapper types for adjusting a type's alignment to increase the number of markable lower
//! bits.
```

**Comment by Amanieu** [comment]
In light of these results, maybe we should go with plan B: only disallow statics with an alignment over the page size. We would also need to disable constant promotion in the compiler if the type of the value being promoted exceeds the target's maximum alignment.

**Comment by hanna-kruppe** [comment]
That won't help with #70143, though. Given that at least 2/3 root regressions seem exclusively for heap allocated data, maybe we could disallow statics *and locals* with alignment >4K?

**Comment by Amanieu** [comment]
I think that #70143 should be fixed separately in LLVM, which we somewhat control. The problem with statics is that the bug is arguably in the OS and we have no control over that.

**Comment by nikomatsakis** [comment]
We discussed this in our @rust-lang/lang [meeting today](https://github.com/rust-lang/lang-team/blob/master/minutes/2020-04-02.md). We didn't reach a final conclusion but sketched out some of the design space, and we wanted to solicit feedback. Let me try to summarize the issue and options based on our discussion from the meeting.

### The problem

There are two related but independent problems, both related to types with alignment greater than the page size:

* Some dynamic linkers in particular choose not to honor alignment requests for static variables (https://github.com/rust-lang/rust/issues/70022).
* Local variable alignment may not be respected by LLVM (https://github.com/rust-lang/rust/issues/70143).

The second issue in particular [can be considered a bug in LLVM](https://github.com/rust-lang/rust/pull/70193#issuecomment-607729588) and might be fixed on that side.

### One proposed solution: limit the size of all types

This PR as currently implemented takes a simple approach: it forbids any type from having an alignment greater than some target-defined limit (i.e., on linux, the page size). This is backwards incompatible change as we currently put no limit; moreover, limits typically *are* respected, it's only in the case of dynamic linking statics and locals that we have issues. 

### Where do we get regressions?

A crater run was recently performed with [3 root regressions](https://github.com/rust-lang/rust/pull/70193#issuecomment-607504924). These all have to do with heap allocation (afaik?), and hence could in principle be covered by the newer, recently stabilized allocation APIs, which permit one to specify alignment.

We also have an example from ARM embedded (https://github.com/rust-lang/rust/issues/42960) where a higher alignment was desired, but this case concerns static linking so in principle the target description might have a higher limit in this case and hence avoid the error.

There is some question as to the "significance" of those 3 root regressions. On the one hand, the absolute number is small. On the other, we don't know about unknown code, of course, and further, this is a niche feature, and the set of folks using high alignments even more niche, so perhaps 3 regressions indicates that a non-trivial fraction of folks using alignments might be impacted. 

### Another proposed solution: limit the size of types used on statics

We could take a narrower approach and only limit the size of types used on statics. This would correct for all the regressions above, as those types are never used on statics. However, it has some complications, because it means we have to issue the errors "post monomorphization". In particular, when you have implicit constant promotion, those constants may be generic (and in the future we may add generic statics), and hence we have to wait until we know the precise types involved to know their true alignments.

There is a particularly "bad-ish" interaction here in that constant promotion happens without the user's "consent", so to speak -- so effectively the compiler is taking a constant expression that was not written in a static, promoting it to a static, and then erroring out because of the alignment limits on that static.

Also, this approach doesn't fix the case for local variables, but the hope is that we can fix that via LLVM (or else impose similar limits in that case).

### To summarize

- Proposal: Disallow alignment on the types themselves (triggers regressions)
    - Pro: simpler overall
    - Con: triggers known regressions in existing code
        - absolute number of regressions is small
        - but as a fraction of users, maybe significant
-  Proposal: Allow them on types but disallow in static constants
    - Pro: as narrowly tailored as possible to the existing problem
    - Con: triggers post monormorphization errors

### Feedback?

Obviously, please let me know if I made any factual errors or overlooked something in this summary!

In the meeting, we said we'd like feedback from the authors of the crates, so cc @oliver-giersch, @wangrunji0408, @tylerhawkes.

We also hoped for feedback from embedded folks -- so I'll cc @jamesmunns and @japaric and @therealprof and hope they can cc other folks as appropriate.

**Comment by nagisa** [comment]
Generally embedded use cases will utilize a linker script of some sort and the linker scripts, among other things, enable developers to arbitrarily align symbols/sections regardless of what is specified in the source code.

To that end I don’t believe there's a huge value proposition for allowing hugely aligned values at the object level. There is still some, though.

And since just slapping an `#[align(8k)]` is going to be an obvious first thing people try, the error messages should be accompanied with links to a long form explanation document (possibly an issue on github) with alternatives for people to use.

**Comment by oliver-giersch** [comment]
Over-aligning is useful for lock-free algorithms, as it guarantees a certain number of bits in each pointer to an instance of an over-aligned type are not used by the pointer itself and can then be used to store a tag value. To my knowledge, this practically always done on heap-allocated values, so if there were suitable replacements, that could be dealt with (BTW which are those recently stabilized allocation APIs?).

It seems slightly unfortunate though. It was rather annoying in pre-C++17 to manually override `new` and `delete` for a class and use `aligned_alloc`, whereas now you can simply use `alignas` on a type and expect allocations to respect it. Another nice feature of C++ that Rust lacks is using valid constant expressions as alignment specifiers like so:

```cpp
template <size_t N>
struct alignas(N) aligned_type { ... };
```
(that's a tangent, though)

**Comment by korken89** [comment]
Hi, I'm part of the embedded WG and when I was experimenting with flash writers some time ago I used this feature up to 128k (which was the largest flash page of the chip in question STM32F405). 
As @nagisa says we can use linker scripts, but having linker script extensions for specific variables is not ergonomic at all. 
There should be some other workaround IMO then to cater for the embedded use cases. We are quite fond on weird alignments that the hardware imposes on us.

**Comment by jamesmunns** [comment]
I generally agree with @nagisa, my first thought was that the idiomatic way to place specific statics at addresses (not too uncommon on bare metal systems) would be using `#[link_section = "..."]` and a linker script - at least for the linked error report.

That being said, there are other cases I could see where having this at the type level could be useful (if you had a bunch of structures that need to each be aligned, like for DMA descriptors). I think @korken89 covered this a bit.

I'll need to review what is motivating the shrink/not generally covering a full `u32` for alignment options.

**Comment by nikomatsakis** [comment]
During the meeting, @Amanieu was suggesting that -- for embedded use cases -- the "per target" limit could be quite high, since dynamic linkers are not an issue. In other words, high alignments might still "just work", if we set things up properly. @Amanieu does that sound right?

**Comment by Amanieu** [comment]
What I specifically meant is that this limitation only really applies to position-independent code since that relies on the program loader (i.e. the OS) to respect the ELF alignment constraints when choosing an address for the program or shared library. This is not an issue for non-PIC code since the linker can choose a suitably aligned base address.

Note that on Windows the situation is slightly different since the object file format itself doesn't support alignments larger than 8K, so there's no way to tell the linker about our alignment constraints.

**Comment by tylerhawkes** [comment]
@nikomatsakis @Amanieu 
I'm the author of maligned, one of the crater regressions. I work on an aarch64 based camera where we need to copy megabytes off the dsp and it has minimum alignment requirements and must be written to . We definitely don't need anything even as high as 4k, but when I published to crates.io 1 Mb seemed like a good stopping point because macros made it so easy to make the different alignment structs. In maligned each of the alignment structs above 128k are feature gated. There are two ways that we use the project internally. The first is by creating aligned byte arrays on the stack. Each of the structs that implement Alignment are also the same size as the alignment because they hold a byte array which can be used as a buffer on the stack that is guaranteed to be aligned and can speed up memcpy calls, so I'm guessing this use case is affected by the llvm bug mentioned above? 

The second way is to use the structs as a generic parameter to other structs to change the alignment per use case by including something like 
```
struct NeedsAlignment<A: maligned::Alignment> {
    ... some fields ...
    _aligned: [A; 0]
}
```
We do this when we treat cma allocated memory from the kernel as the heap since we get it in big chunks and may need to break it up and we have a struct with a `#[repr(C)]` layout that we use for reading from and writing to the cma memory.

I would definitely like the ability to have each binary choose its maximum alignment and if that is the case it would be nice to have a cfg like `target_pointer_width` that could be checked to ensure that even if a feature is enabled that the structs with alignments that are too big are disabled. 

Just choosing 4k is ok but the linux kernel has support for different page alignments up to at least 64k so even when compiling to the the `aarch64-unknown-linux-gnu` target it would be nice to have the ability to increase it if someone is running a custom kernel since aarch64 is supposed to support up to 64k page sizes (I'm no expert on this though).

**Comment by nikomatsakis** [comment]
Thanks @tylerhawkes, that's useful information

**Comment by joshtriplett** [comment]
Discussions about this in the lang team meetings didn't reach a clear conclusion; we would like to catch issues where people won't get the alignment they expect, but we don't want to break code that knows what it's doing (static linking, linker scripts, etc).

I have a proposal to move forward with this: why don't we establish a (target-specific) threshold for alignment that's guaranteed to work, and above that threshold, we warn by default? The warning message can explain that alignment above that threshold may not always work on the target platform, give them some general guidance, and tell the user that if they know it will work for their use case, they can disable the lint. And if we find a case where our lint doesn't match the properties of the target platform, we can fix the lint, but meanwhile we haven't broken people's code.

Thoughts?

**Comment by Mark-Simulacrum** [comment]
Question: How would we expect unsafe code to interact with such a platform-specific limit? e.g. would we expect `[T]::align_to` to use the higher-than-we-guarantee alignment?

Basically I think it boils down to whether `mem::align{,_of}` and similar functions would return the specified alignment or lie.

**Comment by nagisa** [comment]
> I have a proposal to move forward with this: why don't we establish a (target-specific) threshold for alignment that's guaranteed to work, and above that threshold, we warn by default?

I think it would be fine to warn for the common denominator between platforms (i.e. 4k) just to facilitate portable code, and people who know they are not targeting any of the affected platforms can just `allow` the lint.

**Comment by Amanieu** [comment]
Considering the fact that this is a soundness issue (it allows you to trigger UB with only safe code), I think that we should make this at the very least a deny-by-default lint.

However I feel somewhat uncomfortable patching over a soundness issue with a lint (I don't believe we have a precedent for that) so we might also want to say that this will become a hard error in the future. We should re-evaluate the situation after a few warning cycles to see if large alignments are still in use.

**Comment by nagisa** [comment]
We need not allow alignments on platforms where they cannot work. The two
approaches can be combined - the warn lint only informs user of
compatibility concerns and if they are compiling for an affected platform
we fail properly?

**Comment by nikomatsakis** [comment]
@Amanieu 

> Considering the fact that this is a soundness issue (it allows you to trigger UB with only safe code), I think that we should make this at the very least a deny-by-default lint.

Can you say a bit more about the details of the soundness bug? It seems like it could come from two places: (a) folks making assumptions in unsafe code (potentially addressed by the lint) and (b) our code generation making assumptions. I would think that for the latter, we would basically be trying to alter things to not assume that 'very large' alignments are true, not sure if that's plausible or not.

I think that, for me personally, I'd prefer to have the alignments be always accurate and to use a post-monomorphization error to enforce them, than to have them be "sometimes enforced" and use a lint.

I also think the lint isn't really an option if it still leaves an unsoundness hole.

(Side note, I feel like this whole discussion would benefit slightly if we had a kind of 'comprehensive listing' of what kinds of considerations are checked and when that we could reference. That doesn't seem *too hard* to try and create, even. Hmm.)

**Comment by Amanieu** [comment]
The main problem with the post-monomorphization error is that it is dependent on constant promotion: a value with an oversized alignment cannot be promoted to the constant, but we can't detect that since types are not yet known when constant promotion is performed.

**Comment by nikomatsakis** [comment]
@Amanieu Ah yes, thank you! I had forgotten that wrinkle. That actually strikes me as a pretty serious concern.

**Comment by nikomatsakis** [comment]
Er, well, I guess it was in my summary, but it wasn't highlighted in the "list of cons" as much as I'd like. Corrected.

File: src/librustc_attr/builtin.rs [hunk 3]
```diff
@@ -908,11 +908,12 @@ pub fn find_repr_attrs(sess: &ParseSess, attr: &Attribute) -> Vec<ReprAttr> {
                     let parse_alignment = |node: &ast::LitKind| -> Result<u32, &'static str> {
                         if let ast::LitKind::Int(literal, ast::LitIntType::Unsuffixed) = node {
                             if literal.is_power_of_two() {
-                                // rustc::ty::layout::Align restricts align to <= 2^29
-                                if *literal <= 1 << 29 {
+                                // Many targets don't support global variables
+                                // with alignment greater than the page size.
```
Comment (RalfJung) [inline src/librustc_attr/builtin.rs:912]: And also page size is always at last 4096, I take it? That is left implicit here, right now the comment doesn't entirely match what the code does.

**Comment by RalfJung** [comment]
> Can you say a bit more about the details of the soundness bug?

https://github.com/rust-lang/rust/issues/70022 can cause UB in safe code, because we can create insufficiently aligned references.

**Comment by dtolnay** [comment]
I am not the right person to review this. Please assign someone from the lang team until consensus is reached, and then possibly compiler team if needed for the code review.

r? @Dylan-DPC

**Comment by Amanieu** [comment]
I believe the latest consensus is that we should emit a warning, with the intention of changing it to a hard error later. I need to update the PR to implement that.

**Comment by RalfJung** [comment]
:rofl: I support this change but I am not a lang team member and I'm mostly unfamiliar with the code involved... I'll see if I can find someone.

**Comment by RalfJung** [comment]
Quoting @joshtriplett  from Zulip:

> I can confirm that [that was the decision](https://github.com/rust-lang/rust/pull/70193#issuecomment-625946957). We did, however, want the limit to come from the target definition, so that targets can easily define a larger limit.

If targets can override this, we need to find a good way to make sure people do not think that unsafe code can rely on 4k being the maximum alignment. (Some people seem to already think that now, for some reason.)

**Comment by Amanieu** [comment]
If you look at the [list](https://en.wikipedia.org/wiki/Page_(computer_memory)#Multiple_page_sizes) of pages sizes per archtecture, you'll see that all architectures have 4KB as their smallest page size. So I don't think it really makes sense to make this a per-target attribute.

**Comment by chorman0773** [comment]
With respect to 65816, I specify for C/++ that the maximum alignment of any
type is 32k. It doesn't really make sense to impose something stricter on
statics in rust. I would want a per-target attribute, if only so I can
support the same basic limitations in all languages I support.

**Comment by nikomatsakis** [comment]
@Amanieu I believe part of the motivation here was that the limitation of 4K only arises because of interactions with the dynamic linker on some systems (and some LLVM limitations, I believe), and the hypothesis was that on some embedded platforms those things would be non-issues (and there would be legitimate reasons to want higher alignment requirements). It does seem like 'per target' may not be the most direct way to reason about those limits though (e.g., I may be on some target that supports dynamic linking but I may not plan to use it).

**Comment by Amanieu** [comment]
How about this: alignments greater than 4K are not allowed when generating PIC (position independent) code.

**Comment by RalfJung** [comment]
When compiling an rlib, I don't think we know if that code will be position-independent or not -- or do we?

**Comment by hanna-kruppe** [comment]
At least as long as rlibs can contain machine code and not just MIR, we (need to) know. But if this is a monomorphization time check (I think it might needs to be, due to promotion in generic functions?), that won't help all code, nor will it fire in `check` builds.

**Comment by nikomatsakis** [comment]
It is not currently a monomorphization time check, though that was proposed as a possible alternative.

**Comment by Dylan-DPC** [comment]
@Amanieu what's the status of this?

**Comment by Dylan-DPC** [comment]
Closing this after a discussion with the author.

**Pull request closed without merging**